go/tls/tls_test.go

```diff
@@ -0,0 +1,342 @@
+func newUint8(n uint8) *uint8    { return &n }
+func newUint16(n uint16) *uint16 { return &n }
+func newUint24(n Uint24) *Uint24 { return &n }
+func newUint32(n uint32) *uint32 { return &n }
+func newUint64(n uint64) *uint64 { return &n }
+func newInt16(n int16) *int16    { return &n }
+func newEnum(n Enum) *Enum       { return &n }
+
+func TestUnmarshalMarshalWithParamsRoundTrip(t *testing.T) {
+	var tests = []struct {
+		data   string // hex encoded
+		params string
+		item   interface{}
+	}{
+		{"00", "", newUint8(0)},
+		{"03", "", newByte(3)},
+		{"0101", "", newUint16(0x0101)},
+		{"010203", "", newUint24(0x010203)},
+		{"000000", "", newUint24(0x00)},
+		{"00000009", "", newUint32(0x09)},
+		{"0000000901020304", "", newUint64(0x0901020304)},
+		{"030405", "", &[3]byte{3, 4, 5}},
+		{"03", "", &[1]byte{3}},
+		{"0001", "size:2", newEnum(1)},
+		{"0100000001", "size:5", newEnum(0x100000001)},
+		{"12", "maxval:18", newEnum(18)},
+		// Note that maxval is just used to give enum size; it's not policed
+		{"20", "maxval:18", newEnum(32)},
+		{"020a0b", "minlen:1,maxlen:5", &[]byte{0xa, 0xb}},
+		{"020a0b0101010203040011", "", &testStruct{Data: []byte{0xa, 0xb}, IntVal: 0x101, Other: [4]byte{1, 2, 3, 4}, Enum: 17}},
+		{"000102", "", &testVariant{Which: 0, Val16: newUint16(0x0102)}},
+		{"0101020304", "", &testVariant{Which: 1, Val32: newUint32(0x01020304)}},
+		{"0001020104030201", "", &testTwoVariants{Which: 0, Val16: newUint16(0x0102), Second: 1, Second32: newUint32(0x04030201)}},
+		{"06010102020303", "", &testNonByteSlice{Vals: []uint16{0x101, 0x202, 0x303}}},
+		{"00", "", &testSliceOfStructs{Vals: []testVariant{}}},
+		{"080001020101020304", "",
+			&testSliceOfStructs{
+				Vals: []testVariant{
+					testVariant{Which: 0, Val16: newUint16(0x0102)},
+					testVariant{Which: 1, Val32: newUint32(0x01020304)},
+				},
+			},
+		},
+		{"000a00030102030003040506", "",
+			&testSliceOfSlices{
+				Inners: []testInnerType{
+					testInnerType{Val: []byte{1, 2, 3}},
+					testInnerType{Val: []byte{4, 5, 6}},
+				},
+			},
+		},
+		{"011011", "", &testAliasEnum{Val: 1, Val16: newUint16(0x1011)}},
+		{"0403", "", &SignatureAndHashAlgorithm{Hash: SHA256, Signature: ECDSA}},
+		{"04030003010203", "",
+			&DigitallySigned{
+				Algorithm: SignatureAndHashAlgorithm{Hash: SHA256, Signature: ECDSA},
+				Signature: []byte{1, 2, 3},
+			},
+		},
+	}
+	for _, test := range tests {
+		inVal := reflect.ValueOf(test.item).Elem()
+		pv := reflect.New(reflect.TypeOf(test.item).Elem())
+		val := pv.Interface()
+		inData, _ := hex.DecodeString(test.data)
+		if _, err := UnmarshalWithParams(inData, val, test.params); err != nil {
+			t.Errorf("Unmarshal(%s)=nil,%q; want %+v", test.data, err.Error(), inVal)
+		} else if !reflect.DeepEqual(val, test.item) {
+			t.Errorf("Unmarshal(%s)=%+v,nil; want %+v", test.data, reflect.ValueOf(val).Elem(), inVal)
+		}
+
+		data, err := MarshalWithParams(inVal.Interface(), test.params)
```

Comment:
I believe you can put this line in the below if statement, and the else-if condition will still be able to access the value of data.